src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java

```diff
@@ -251,6 +251,20 @@ static void verifySucceededCleanly(Jenkins j, WorkflowRun run) throws Exception
         assertHasTimingAction(run.getExecution());
     }
 
+    /** A minimal check of resuming the job with no node checks, in case of a dirty shutdown.
+     * Currently dirty shutdowns have the potential to lose the head node ID if they happen at the moment of writing
+     * it to build.xml
+     */
+    static void verifyDirtyResumed(JenkinsRule rule, WorkflowRun run, String logStart) throws Exception {
+        assert run.isBuilding();
```

Comment:
I know you copied this from other tests, but IMO it is very strange and potentially confusing to use Java `assert` statements in tests:

```suggestion
        Assert.assertTrue(run.isBuilding());
```

Also though, it looks like this assertion is potentially timing sensitive (maybe the build finishes very quickly). Maybe better to just delete it and not check anything before `rule.waitForCompletion`.

Reply:
> because it always does so on a waiting step like sleep/semaphore (every pipeline restart seems to be meant to happen on a sleep), and it can't successfully finish faster than the allotted sleep time

Unless we use something like the `semaphore` step to _guarantee_ that the build is not allowed to finish before we check `run.isBuilding`, I think there is a potential timing issue, even if it is very unlikely.